src/sympc/session/session.py

```diff
@@ -1,40 +1,142 @@
+import operator
+
 from uuid import uuid1
-from ..config import Config
-from copy import deepcopy
+from typing import Union
+from typing import List
+from typing import Any
+
+from sympc.config import Config
+from sympc.session.utils import get_type_from_ring
+from sympc.session.utils import generate_random_element
+
+import secrets
+
+
+# TODO: This should not be here
+import torch
 
 
 class Session:
-    def __init__(self, config=None, parties=None, ttp=None, uuid=None):
+    NOT_COMPARE = {"id", "description", "tags"}
+    __slots__ = {
+        # Populated in Syft
+        "id",
+        "tags",
+        "description",
+        "uuid",
+        "parties",
+        "trusted_third_party",
+        "crypto_store",
+        "protocol",
+        "config",
+        "przs_generators",
+        "rank",
+        "session_ptr",
+        "ring_size",
+        "tensor_type",
+        "min_value",
+        "max_value",
+    }
+
+    def __init__(
+        self,
+        parties: Union[List[Any], None] = None,
+        ring_size: int = 2 ** 64,
+        config: Union[None, Config] = None,
+        ttp: Union[None] = None,
+        uuid: Union[None, uuid1] = None,
+    ) -> None:
+
         self.uuid = uuid1() if uuid is None else uuid
 
         # Each worker will have the rank as the index in the list
+        # Only the party that is the CC (Control Center) will have access
+        # to this
         self.parties = parties
 
         # Some protocols require a trusted third party
         # Ex: SPDZ
         self.trusted_third_party = ttp
+
         self.crypto_store = {}
         self.protocol = None
         self.config = config if config else Config()
 
+        self.przs_generators = None
+
         # Those will be populated in the setup_mpc
-        self.rank = None
+        self.rank = -1
         self.session_ptr = []
 
-    def get_copy(self):
-        session_copy = Session()
-        session_copy.uuid = deepcopy(self.uuid)
-        session_copy.parties = list(self.parties)
-        session_copy.trusted_third_party = self.trusted_third_party
-        session_copy.crypto_store = {}
-        session_copy.protocol = self.protocol
-        session_copy.config = deepcopy(self.config)
-        session_copy.rank = self.rank
-        session_copy.session_ptr = list(self.session_ptr)
-
-        return session_copy
-
-    def setup_mpc(self):
-        raise NotImplementedError(
-            "Need to be able to send the session to other parties"
+        # Ring size
+        self.tensor_type = get_type_from_ring(ring_size)
+        self.ring_size = ring_size
+        self.min_value = -(ring_size) // 2
+        self.max_value = (ring_size - 1) // 2
+
+    def przs_generate_random_share(
+        self, shape: Union[tuple, torch.Size], generators: List[torch.Generator]
+    ) -> "ShareTensor":
+
+        from sympc.tensor import ShareTensor
+
+        gen0, gen1 = generators
```

Comment:
Those generators are not that secure because we kinda use a "seed" value to initialize them (that seed is - for the moment - not even generated by them).

In a real scenario, I (let's say party 1) will generate a seed and send it to you Theo, and you will do the same with a seed you generated.